<a href="https://colab.research.google.com/github/kunal24bit/NLP/blob/main/Tokenization_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Tokenization**

Tokenization is the process of breaking up the original text in to component pieces(Tokens). Tokens are the basic building blocks of a doc object. Everything that helps us understand the meaning of text is dervied from tokens and their relationship to one another.

Prefix: Character(s) at the begining.
Suffix: Character(s) at the end.
Infix:  Character(s) in between

In [3]:
import spacy

nlp = spacy.load("en_core_web_sm")

In [4]:
mystring = '"We\'re moving to L.A!"'

In [5]:
print(mystring)

"We're moving to L.A!"


In [6]:
doc = nlp(mystring)

In [7]:
for token in doc:
  print(token.text)

"
We
're
moving
to
L.A
!
"


In [19]:
doc2 =  nlp(u"We\'re here to help! Send snail-mail, email supoort@oursite.com or visit us at http://oursite.com")

In [9]:
for t in doc2:
  print(t.text, t.pos_)

We PRON
're AUX
here ADV
to PART
help VERB
! PUNCT
Send VERB
snail NOUN
- PUNCT
mail NOUN
, PUNCT
email NOUN
supoort@oursite.com X
or CCONJ
visit VERB
us PRON
at ADP
http://oursite.com X


In [10]:
doc3 = nlp(u"A 5km NYC can ride cost $10.30")

In [11]:
for t in doc3:
  print(t)

A
5
km
NYC
can
ride
cost
$
10.30


In [12]:
doc4 = nlp(u"Let's visit St. Louis in the U.S. next year")

In [13]:
for t in doc4:
  print(t)

Let
's
visit
St.
Louis
in
the
U.S.
next
year


In [14]:
#for calculating the of doc
len(doc4)

10

In [16]:
doc4.vocab

In [17]:
len(doc4.vocab)

510

So there are 510 different tokens in spacy vocab.

In [21]:
doc5 = nlp(u"It is better to give than recieve")

In [22]:
for t in doc5:
  print(t)

It
is
better
to
give
than
recieve


In [23]:
doc5[0]

It

In [24]:
doc5[2:5]

better to give

**Tokens cannot be reassign.**

In [25]:
doc5[0] = 'test'

TypeError: ignored

**Spacy can understand named entities**

In [35]:
doc8 = nlp(u"Apple to build a Hong Kong factory for $6 million")

In [36]:
for token in doc8:
  print(token.text, end ='  |  ')

Apple  |  to  |  build  |  a  |  Hong  |  Kong  |  factory  |  for  |  $  |  6  |  million  |  

In [41]:
for entity in doc8.ents:
  print(entity)
  print(entity.label_) # entity.label_ tells us that Apple is an Organization, Hong Kong is city and $6 million is money.
  print(str(spacy.explain(entity.label_)))
  print('\n')

Apple
ORG
Companies, agencies, institutions, etc.


Hong Kong
GPE
Countries, cities, states


$6 million
MONEY
Monetary values, including unit




In [42]:
doc9 = nlp(u"Autonomous cars shift insurance liability towards manufacturers.")

In [43]:
for chunk in doc9.noun_chunks:
  print(chunk)

Autonomous cars
insurance liability
manufacturers
